lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs

```diff
@@ -280,51 +356,74 @@ _decodeSignedTx bytes = do
 
 _minimumFee
     :: FeePolicy
-    -> WithDelegation
+    -> [Certificate]
     -> CoinSelection
     -> Fee
-_minimumFee policy withDelegation coinSel =
-    computeFee $ computeTxSize withDelegation coinSel
+_minimumFee policy certs coinSel =
+    computeFee $ computeTxSize certs coinSel
   where
     computeFee :: Integer -> Fee
     computeFee size =
-        Fee $ ceiling (a + b*fromIntegral size)
+        Fee $ ceiling (a + b*fromIntegral size + if needsDeposit then c else 0)
       where
-        LinearFee (Quantity a) (Quantity b) (Quantity _unused) = policy
+        LinearFee (Quantity a) (Quantity b) (Quantity c) = policy
+        needsDeposit = isJust $ find (== KeyRegistrationCertificate) certs
+
+realFee :: [(TxIn, TxOut)] -> [TxOut] -> Fee
+realFee inps outs = Fee
+    $ sum (map (getCoin . coin . snd) inps)
+    - sum (map (getCoin . coin) outs)
 
 computeTxSize
-    :: WithDelegation
+    :: [Certificate]
     -> CoinSelection
     -> Integer
-computeTxSize (WithDelegation withDelegation) (CoinSelection inps outs chngs) =
+computeTxSize certs (CoinSelection inps outs chngs) =
     SL.txsize $ SL.Tx unsigned wits metadata
  where
     metadata = SL.SNothing
 
-    unsigned = mkUnsignedTx maxBound inps outs'
-        $ if withDelegation then [dummyRegisterCert, dummyDelegateCert] else []
+    unsigned = mkUnsignedTx maxBound inps outs' certs' (realFee inps outs')
       where
         outs' :: [TxOut]
         outs' = outs <> (dummyOutput <$> chngs)
 
         dummyOutput :: Coin -> TxOut
-        dummyOutput = TxOut $ Address $ BS.pack (1:replicate 64 0)
-
-        dummyKeyHash = SL.KeyHash . Hash.UnsafeHash $ BS.pack (1:replicate 64 0)
-        dummyRegisterCert = Cardano.shelleyRegisterStakingAddress dummyKeyHash
-        dummyDelegateCert = Cardano.shelleyDelegateStake dummyKeyHash dummyKeyHash
-
-    addrWits = Set.map dummyWitness $ Set.fromList (fst <$> inps)
+        dummyOutput = TxOut $ Address $ BS.pack (1:replicate 56 0)
+
+        dummyKeyHash = SL.KeyHash . Hash.UnsafeHash $ BS.pack (replicate 28 0)
+
+        certs' = flip map certs $ \case
+            PoolDelegationCertificate ->
+                Cardano.shelleyDelegateStake dummyKeyHash dummyKeyHash
+            KeyRegistrationCertificate ->
+                Cardano.shelleyRegisterStakingAddress dummyKeyHash
+            KeyDeRegistrationCertificate ->
+                Cardano.shelleyDeregisterStakingAddress dummyKeyHash
+
+    (addrWits, certWits) =
+        ( Set.map dummyWitness $ Set.fromList (fst <$> inps)
+        , if null certs
+            then Set.empty
+            else Set.singleton (dummyWitness dummyTxIn)
+        )
```

Comment:
Had to revise the size estimation too because it was wrong :/ ... 

a) The hash length were still 32 bytes 
b) Certificate witnesses weren't counted

Turned out that, until I introduced the deposit amount, b) and a) were both balancing each other (we were allocating too much fee because of the wrong hash sizes, and it was coping with the lack of witness for the certificate :shrug: )